File: 3/part_two.py
#!/bin/env python3
import re

with open('input') as f:
    data = f.read()

r = r'(do\(\))|(don\'t\(\))|mul\((\d+),(\d+)\)'

res = re.findall(r, data)
#print(res)

res2 = [list(filter(None, x)) for x in res]

mult = True
total = 0

for el in res2:
    if len(el) == 1:
        mult = el[0] == "do()"
        continue
    if mult:
        total += int(el[0]) * int(el[1])

print(total)

File: 3/part_two.hs
import Data.Char (isDigit)
import Text.ParserCombinators.ReadP

data Mul = Mul {x :: Int, tot :: Int, on :: Bool} deriving (Show)

getMul :: ReadP Mul
getMul = do
  string "mul("
  d1 <- fmap read $ many1 $ satisfy isDigit
  string ","
  d2 <- fmap read $ many1 $ satisfy isDigit
  string ")"
  return (Mul (d1 * d2) 0 False)

getDo = do
  string "do()"
  return (Mul 0 0 True)

getDont = do
  string "don't()"
  return (Mul 0 0 False)

sumMul nv tt
  | x nv == 0 = Mul 0 (tot tt) (on nv)
  | otherwise = Mul 0 (if on tt then x nv + tot tt else tot tt) (on tt)

parsePart2 [] total = total
parsePart2 content total =
  case readP_to_S (getMul <++ getDont <++ getDo) content of
    [] -> parsePart2 (tail content) total
    [(cur, rest)] -> parsePart2 rest $ sumMul cur total
    _ -> error "Oww"

main = do
  content <- readFile "../data/2024/3/input"
  let resp = parsePart2 content (Mul 0 0 True)
  print (tot resp)
